File: juno/strategies/double_ma.py
from __future__ import annotations

import operator
from dataclasses import dataclass

from juno import Advice, Candle, CandleMeta, indicators
from juno.constraints import Int, Pair
from juno.indicators import MA, Ema
from juno.typing import Constructor
from juno.utils import get_module_type

from .strategy import Signal, Strategy, ma_choices


@dataclass
class DoubleMAParams(Constructor):
    short_ma: str = "ema"
    long_ma: str = "ema"
    short_period: int = 5  # Common 5 or 10. Daily.
    long_period: int = 20  # Common 20 or 50.

    def construct(self) -> DoubleMA:
        return DoubleMA(
            short_ma=self.short_ma,
            long_ma=self.long_ma,
            short_period=self.short_period,
            long_period=self.long_period,
        )


# Signals long when shorter average crosses above the longer.
# Signals short when shorter average crosses below the longer.
# J. Murphy 203
class DoubleMA(Signal):
    @staticmethod
    def meta() -> Strategy.Meta:
        return Strategy.Meta(
            constraints={
                "short_ma": ma_choices,
                "long_ma": ma_choices,
                ("short_period", "long_period"): Pair(Int(1, 100), operator.lt, Int(2, 101)),
            }
        )

    _short_ma: MA
    _long_ma: MA
    _advice: Advice = Advice.NONE

    def __init__(
        self,
        short_ma: str = Ema.__name__.lower(),
        long_ma: str = Ema.__name__.lower(),
        short_period: int = 5,  # Common 5 or 10. Daily.
        long_period: int = 20,  # Common 20 or 50.
    ) -> None:
        assert short_period > 0
        assert short_period < long_period

        self._short_ma = get_module_type(indicators, short_ma)(short_period)
        self._long_ma = get_module_type(indicators, long_ma)(long_period)

    @property
    def advice(self) -> Advice:
        return self._advice

    @property
    def maturity(self) -> int:
        return max(self._long_ma.maturity, self._short_ma.maturity)

    @property
    def mature(self) -> bool:
        return self._long_ma.mature and self._short_ma.mature

    def update(self, candle: Candle, _: CandleMeta) -> None:
        self._short_ma.update(candle.close)
        self._long_ma.update(candle.close)

        if self._long_ma.mature and self._short_ma.mature:
            if self._short_ma.value >= self._long_ma.value:
                self._advice = Advice.LONG
            elif self._short_ma.value < self._long_ma.value:
                self._advice = Advice.SHORT

File: juno_rs/src/tactics/double_ma.rs
use super::{Signal, Tactic};
use crate::{
    genetics::Chromosome,
    indicators::{ma_from_adler32, MA, MA_CHOICES},
    Advice, Candle,
};
use juno_derive_rs::*;
use rand::prelude::*;

#[derive(Chromosome, Clone, Debug)]
#[repr(C)]
pub struct DoubleMAParams {
    pub short_ma: u32,
    pub long_ma: u32,
    pub periods: (u32, u32),
}

fn short_ma(rng: &mut StdRng) -> u32 {
    MA_CHOICES[rng.gen_range(0, MA_CHOICES.len())]
}
fn long_ma(rng: &mut StdRng) -> u32 {
    MA_CHOICES[rng.gen_range(0, MA_CHOICES.len())]
}
fn periods(rng: &mut StdRng) -> (u32, u32) {
    loop {
        let (s, l) = (rng.gen_range(1, 100), rng.gen_range(2, 101));
        if s < l {
            return (s, l);
        }
    }
}

pub struct DoubleMA {
    short_ma: Box<dyn MA>,
    long_ma: Box<dyn MA>,
    advice: Advice,
}

unsafe impl Send for DoubleMA {}
unsafe impl Sync for DoubleMA {}

impl Tactic for DoubleMA {
    type Params = DoubleMAParams;

    fn new(params: &Self::Params) -> Self {
        let (short_period, long_period) = params.periods;
        assert!(short_period > 0);
        assert!(short_period < long_period);

        Self {
            short_ma: ma_from_adler32(params.short_ma, short_period),
            long_ma: ma_from_adler32(params.long_ma, long_period),
            advice: Advice::None,
        }
    }

    fn maturity(&self) -> u32 {
        self.long_ma.maturity()
    }

    fn mature(&self) -> bool {
        self.long_ma.mature()
    }

    fn update(&mut self, candle: &Candle) {
        self.short_ma.update(candle.close);
        self.long_ma.update(candle.close);

        if self.long_ma.mature() {
            if self.short_ma.value() > self.long_ma.value() {
                self.advice = Advice::Long;
            } else if self.short_ma.value() < self.long_ma.value() {
                self.advice = Advice::Short;
            }
        }
    }
}

impl Signal for DoubleMA {
    fn advice(&self) -> Advice {
        self.advice
    }
}
